### ISSUE TITLE
Show Symbols

### ISSUE THREAD

**Issue by archaeron** [hoovercj/vscode-ghc-mod#28]
Hi

I did a quick proof of concept of implementing the `documentSymbolProvider` capability.
You need `fast-tags` installed.
Would you be interesed in this?

Here is my quick-and-dirty version:

```
import * as cp from 'child_process';
let Path = require('path');

function filepathToUri(filepath: string): string
{
    if (!Path.isAbsolute(filepath)) {
        filepath = Path.join(this.workspaceRoot || '', filepath || '');
    }
    return `file:///${filepath.replace('\\', '/')}`;
}

function toSymbolKind(rawKind : string) : SymbolKind
{
    console.log("Raw kind", "'" + rawKind + "'");
    switch(rawKind.trim())
    {
        case 'm':
            return SymbolKind.Module
        case 'f':
            return SymbolKind.Function
        case 'c':
            return SymbolKind.Class
        case 't':
            return SymbolKind.Interface
        case 'C':
            return SymbolKind.Constructor
        case 'o':
            return SymbolKind.Method
        default:
            return SymbolKind.Function
    }
}

export function parseTags(rawTags : String) : SymbolInformation[]
{
    let symbolInformations = rawTags
        .split("\n")
        .slice(1)
        .map((tagLine) => { console.log("Line: ", tagLine); return tagLine.split("\t") })
        .filter((line) => line.length === 4)
        .map(([name, path, line, kind]) =>
            {
                let uri = filepathToUri(path);
                let lineNumber = parseInt(line, 10) - 1;
                let range =
                    Range.create(
                        Position.create(lineNumber, 0),
                        Position.create(lineNumber, 0)
                    );
                let symbolKind = toSymbolKind(kind);
                return <SymbolInformation>{
                    name : name,
                    kind : symbolKind,
                    location: Location.create(uri, range)
                };
            }
        )

    return symbolInformations;
}

export function symbolsForFile(documentSymbolParams : DocumentSymbolParams) : Promise<SymbolInformation[]>
{
    let uri = documentSymbolParams.textDocument.uri;
    return new Promise((resolve, reject) =>
        {
            let command = "fast-tags -o - " + uriToFilePath(uri);
            cp.exec(command, (error, stdout, stderr) =>
                {
                    if(error || stderr)
                    {
                        reject();
                    }
                    resolve(parseTags(stdout.toString("UTF8")));
                }
            )
        }
    )
}

function initializeSymbolProvider(): void {
    connection.onDocumentSymbol(symbolsForFile);
}
```

**Comment by hoovercj** [comment]
@archaeron great idea, thanks for sketching this out! I went ahead and tweaked this to work with the existing extension in PR #29. It's working for me on Windows, are you running linux and want to give it a test? I also added "Show all symbols" support, so if you open a project you should be able to type # and search for a symbol in a different file.

I look forward to your feedback on the PR

**Comment by archaeron** [comment]
@hoovercj awesome!! this is so cool :) I found just one issue. I'll comment on the PR

**Comment by hoovercj** [comment]
Update @archaeron: published this and hasktags support as version 1.0.0

**Comment by archaeron** [comment]
@hoovercj thanks so much! This should make my work so much easyer :)
